Make Aggregation use RowExpression instead

hellium01 opened the pull request:
This will remove the dependency to Expression in following plan nodes:
```
AggregationNode
TableWriterNode
TableFinishNode
```
It is splitted to small commits  that may need to squash together later.

highker commented:
@hellium01 is this ready to review?

hellium01 commented:
Yes, it should be OK to review now.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlannerUtils.java:
```diff
@@ -0,0 +1,43 @@
+            if (sortItem.getNullOrdering() == SortItem.NullOrdering.FIRST) {
+                return SortOrder.ASC_NULLS_FIRST;
+            }
+            else {
```
`else` if redundant

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlannerUtils.java:
```diff
@@ -0,0 +1,43 @@
+            if (sortItem.getNullOrdering() == SortItem.NullOrdering.FIRST) {
+                return SortOrder.DESC_NULLS_FIRST;
+            }
+            else {
```
`else` if redundant

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementFilteredAggregations.java:
```diff
@@ -93,8 +93,8 @@ public Result apply(AggregationNode aggregation, Captures captures, Context cont
             FunctionCall call = entry.getValue().getCall();
```
`call` is still used to create an `Aggregation` down below.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteSpatialPartitioningAggregation.java:
```diff
@@ -44,25 +44,25 @@
 /**
  * Re-writes spatial_partitioning(geometry) aggregations into spatial_partitioning(envelope, partition_count)
  * on top of ST_Envelope(geometry) projection, e.g.
- *
```
Quite a few unrelated changes

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteSpatialPartitioningAggregation.java:
```diff
@@ -71,17 +71,17 @@ public RewriteSpatialPartitioningAggregation(Metadata metadata)
         return aggregation.getAggregations().values().stream()
-                .map(Aggregation::getCall)
-                .anyMatch(call -> call.getName().toString().equals(NAME) && call.getArguments().size() == 1);
+                .anyMatch(a -> metadata.getFunctionManager().getFunctionMetadata(a.getFunctionHandle()).getName().equals(NAME)
```
`a` -> `aggregation`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteSpatialPartitioningAggregation.java:
```diff
@@ -44,25 +44,25 @@
-    private static final Pattern<AggregationNode> PATTERN = aggregation()
-            .matching(RewriteSpatialPartitioningAggregation::hasSpatialPartitioningAggregation);
+    private final Pattern<AggregationNode> pattern = aggregation()
+            .matching(this::hasSpatialPartitioningAggregation);
```
Put this to the previous line

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteSpatialPartitioningAggregation.java:
```diff
@@ -71,17 +71,17 @@ public RewriteSpatialPartitioningAggregation(Metadata metadata)
-                .map(Aggregation::getCall)
-                .anyMatch(call -> call.getName().toString().equals(NAME) && call.getArguments().size() == 1);
+                .anyMatch(a -> metadata.getFunctionManager().getFunctionMetadata(a.getFunctionHandle()).getName().equals(NAME)
+                        && a.getArguments().size() == 1);
```
Put this to the previous line

hellium01 replied:
It will be a little long to fit in one line.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AggregationNodeUtil.java:
```diff
@@ -0,0 +1,54 @@
+
+    public static OrderingScheme toOrderingScheme(List<SortItem> sortItems, Function<Expression, Symbol> translator)
+    {
+        LinkedHashMap<Symbol, SortOrder> orderings = new LinkedHashMap<>();
```
`Map<Symbol, SortOrder> orderings = new LinkedHashMap<>();`

highker commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushAggregationThroughOuterJoin.java:
```diff
@@ -42,10 +44,11 @@
 public class TestPushAggregationThroughOuterJoin
         extends BaseRuleTest
 {
+    private static final FunctionManager TEST_FUNCTION_MANAGER = createTestMetadataManager().getFunctionManager();
```
New line after

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AggregationNodeUtil.java:
```diff
@@ -0,0 +1,54 @@
+    private AggregationNodeUtil()
+    {}
+
+    public static OrderingScheme toOrderingScheme(List<SortItem> sortItems)
```
I would extract `QueryPlanner::sort` to `PlannerUtils` rather than binding with aggregation node.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SymbolsExtractor.java:
```diff
@@ -81,6 +82,15 @@ private SymbolsExtractor() {}
         return ImmutableSet.copyOf(extractAll(expression).stream().map(variable -> new Symbol(variable.getName())).collect(toSet()));
     }
 
+    public static Set<Symbol> extractUnique(AggregationNode.Aggregation aggregation)
```
Move this to `AggregationUtil`. It is too specific for this class.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SymbolsExtractor.java:
```diff
@@ -97,6 +107,16 @@ private SymbolsExtractor() {}
         return builder.build();
     }
 
+    public static List<Symbol> extractAll(AggregationNode.Aggregation aggregation)
```
`extractUnique` should reuse `extractAll` to avoid dup code

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java:
```diff
@@ -153,7 +154,13 @@ public Result apply(AggregationNode aggregationNode, Captures captures, Context
                 Aggregation aggregation = entry.getValue();
-                FunctionCall call = (FunctionCall) rewriter.rewrite(aggregation.getCall(), context);
+                FunctionCall call = new FunctionCall(
+                        aggregation.getCall().getName(),
```
Why don't we use function manager?

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java:
```diff
@@ -153,7 +154,13 @@ public Result apply(AggregationNode aggregationNode, Captures captures, Context
+                FunctionCall call = new FunctionCall(
+                        aggregation.getCall().getName(),
+                        Optional.empty(),
+                        aggregation.getFilter().map(a -> rewriter.rewrite(a, context)),
```
`a` -> `aggregation`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughJoin.java:
```diff
@@ -102,7 +100,9 @@ else if (allAggregationsOn(aggregationNode.getAggregations(), joinNode.getRight(
     private boolean allAggregationsOn(Map<Symbol, AggregationNode.Aggregation> aggregations, List<Symbol> symbols)
     {
-        Set<Symbol> inputs = extractUnique(aggregations.values().stream().map(AggregationNode.Aggregation::getCall).collect(toImmutableList()));
+        Set<Symbol> inputs = aggregations.values().stream().map(SymbolsExtractor::extractUnique)
```
nits
```java
Set<Symbol> inputs = aggregations.values()
    .stream()
    .map(SymbolsExtractor::extractUnique)
    .flatMap(Set::stream)
    .collect(toImmutableSet());

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformQuantifiedComparisonApplyToLateralJoin.java:
```diff
@@ -149,19 +149,19 @@ private PlanNode rewriteQuantifiedApplyNode(ApplyNode node, QuantifiedComparison
                                     new FunctionCall(MIN, outputColumnReferences),
                                     functionManager.lookupFunction(MIN.getSuffix(), outputColumnTypeSignatures),
-                                    Optional.empty()),
+                                    outputColumnReferences, Optional.empty(), Optional.empty(), false, Optional.empty()),
```
one variable a line

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java:
```diff
@@ -359,16 +358,28 @@ public static Step partialInput(Step step)
         public Aggregation(
                 @JsonProperty("call") FunctionCall call,
                 @JsonProperty("functionHandle") FunctionHandle functionHandle,
+                @JsonProperty("arguments") List<Expression> arguments,
```
I don't think you need to move out `arguments`... Otherwise `FunctionCall` is nothing but a name. Keep `arguments` inside `call` can help the refactoring of directly replacing `FunctionCall` with `CallExpression`

hellium01 replied:
Yes, it was temporarily moved out to be able to convert it to RowExpression (just to make it easier to code). In the end argument is part of CallExpression. Once these squashed, it should not be a problem.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java:
```diff
@@ -144,10 +144,10 @@ public String getGraphvizPlan(Session session, Statement statement, Type planTyp
             case LOGICAL:
                 Plan plan = getLogicalPlan(session, statement, parameters, warningCollector);
-                return graphvizLogicalPlan(plan.getRoot(), plan.getTypes(), session);
+                return graphvizLogicalPlan(plan.getRoot(), plan.getTypes(), session, metadata.getFunctionManager());
```
Add a TODO: function manager will be removed after `CallExpression` is introduced in `AggregationNode`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlannerUtils.java:
```diff
@@ -0,0 +1,66 @@
+        Map<Symbol, SortOrder> orderings = new LinkedHashMap<>();
+        for (SortItem item : sortItems) {
+            Symbol symbol = translator.apply(item.getSortKey());
+            // don't override existing keys, i.e. when "ORDER BY a ASC, a DESC" is specified
```
I only see this logic in `QueryPlanner::sort`. Are there other places having this? We'd better merge the logic into a single helper (at least with `QueryPlanner::sort`).

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/StatisticsAggregationPlanner.java:
```diff
@@ -137,9 +138,8 @@ private ColumnStatisticsAggregation createAggregation(String functionName, Symbo
-                        new FunctionCall(QualifiedName.of(functionName), ImmutableList.of(input)),
                         functionHandle,
-                        Optional.empty()),
+                        ImmutableList.of(input), Optional.empty(), Optional.empty(), false, Optional.empty()),
```
One variable a line

hellium01 commented:
@highker / @wenleix , I have reorganized the commits (hopefully makes it easier to review) and addressed the comments. Please let me know if you have more comments.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SymbolsExtractor.java:
```diff
@@ -137,7 +137,7 @@ private SymbolsExtractor() {}
     }
 
-    private static class SymbolBuilderVisitor
+    public static class SymbolBuilderVisitor
```
Keep it private and use `extractAll` interface in this class instead.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneAggregationSourceColumns.java:
```diff
@@ -57,7 +57,7 @@ public Result apply(AggregationNode aggregationNode, Captures captures, Context
     {
         return Streams.concat(
-                SymbolsExtractor.extractUnique(aggregation.getCall()).stream(),
+                extractUnique(aggregation).stream(),
```
I would keep `AggregationNodeUtil.extractUnique` to avoid confusion...

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneCountAggregationOverScalar.java:
```diff
@@ -55,8 +61,7 @@ public Result apply(AggregationNode parent, Captures captures, Context context)
             requireNonNull(aggregation, "aggregation is null");
-            FunctionCall functionCall = aggregation.getCall();
-            if (!"count".equals(functionCall.getName().getSuffix()) || !functionCall.getArguments().isEmpty()) {
+            if (!"count".equals(functionManager.getFunctionMetadata(aggregation.getFunctionHandle()).getName()) || !aggregation.getArguments().isEmpty()) {
```
Any other better way doing this...? cc: @rongrong

highker replied:
Maybe add `isCountFunction` to function resolution.

hellium01 replied:
I think it probably make sense to add standard aggregation functions such as count to standard function resolution, @rongrong ?

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:
```diff
@@ -299,10 +306,14 @@ private static boolean groupsOnAllColumns(AggregationNode node, List<Symbol> col
                     aggregation.getFunctionHandle(),
+                    aggregation.getArguments().stream().map(argument -> inlineSymbols(sourcesSymbolMapping, argument)).collect(toImmutableList()),
+                    aggregation.getFilter().map(filter -> inlineSymbols(sourcesSymbolMapping, filter)),
+                    aggregation.getOrderBy(), // order by symbol should not be inlinable
```
why is that?

hellium01 replied:
Yes, we should inline this as well. Will change it.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedInPredicateToJoin.java:
```diff
@@ -250,17 +248,12 @@ private static JoinNode leftOuterJoin(PlanNodeIdAllocator idAllocator, AssignUni
-                QualifiedName.of("count"),
-                Optional.empty(),
+        return new AggregationNode.Aggregation(
+                functionManager.lookupFunction("count", ImmutableList.of()),
```
Maybe add `countFunction()` to function resolution

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformExistsApplyToLateralNode.java:
```diff
@@ -162,9 +162,8 @@ private PlanNode rewriteToDefaultAggregation(ApplyNode parent, Context context)
-                                        COUNT_CALL,
                                         functionManager.lookupFunction(COUNT, ImmutableList.of()),
-                                        Optional.empty())),
+                                        ImmutableList.of(), Optional.empty(), Optional.empty(), false, Optional.empty())),
```
one variable per line

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AggregationNodeUtil.java:
```diff
@@ -0,0 +1,58 @@
+    public static AggregationNode.Aggregation count(FunctionManager functionManager)
+    {
+        return new AggregationNode.Aggregation(
+                functionManager.lookupFunction("count", ImmutableList.of()),
```
same

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AggregationNodeUtil.java:
```diff
@@ -0,0 +1,58 @@
+    public static List<Symbol> extractAll(AggregationNode.Aggregation aggregation)
+    {
+        ImmutableList.Builder<Symbol> builder = ImmutableList.builder();
+        SymbolsExtractor.SymbolBuilderVisitor visitor = new SymbolsExtractor.SymbolBuilderVisitor();
```
Use interface than visitor

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeMixedDistinctAggregations.java:
```diff
@@ -422,27 +421,39 @@ private AggregationNode createNonDistinctAggregation(
-                                else {
-                                    arguments.add(argument);
-                                }
+                    // Handling for cases when mask symbol appears in non distinct aggregations too
```
Maybe thhe following code looks better
```java
List<Expression> arguments;
if (!duplicatedDistinctSymbol.equals(distinctSymbol) && entry.getValue().getArguments().contains(distinctSymbol.toSymbolReference())) {
    ImmutableList.Builder<Expression> argumentsBuilder = ImmutableList.builder();
    for (Expression argument : aggregation.getArguments()) {
        if (distinctSymbol.toSymbolReference().equals(argument)) {
            argumentsBuilder.add(duplicatedDistinctSymbol.toSymbolReference());
        }
        else {
            argumentsBuilder.add(argument);
        }
    }
    arguments = argumentsBuilder.build();
}
else {
    arguments = entry.getValue().getArguments();
}
aggregations.put(newSymbol, new Aggregation(
        entry.getValue().getFunctionHandle(),
        arguments,
        Optional.empty(),
        Optional.empty(),
        false,
        Optional.empty()));

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java:
```diff
@@ -173,18 +172,18 @@ private PlanNode rewriteScalarAggregation(
-            FunctionCall call = entry.getValue().getCall();
             Symbol symbol = entry.getKey();
-            if (call.getName().equals(COUNT)) {
+            if (functionManager.getFunctionMetadata(entry.getValue().getFunctionHandle()).getName().equalsIgnoreCase("count")) {
```
same

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java:
```diff
@@ -78,6 +81,19 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
+    public OrderingScheme map(OrderingScheme orderingScheme)
+    {
+        return new OrderingScheme(orderingScheme.getOrderBy().stream().map(this::map).collect(toImmutableList()),
+                orderingScheme.getOrderings().entrySet().stream().collect(toMap(entry -> map(entry.getKey()), entry -> entry.getValue())));
```
Is this the same before and after?

hellium01 replied:
The key get translated here?

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:
```diff
@@ -486,17 +486,23 @@ public Void visitAggregation(AggregationNode node, Void context)
             return processChildren(node, context);
         }
 
+        private String printAggregation(AggregationNode.Aggregation aggregation)
```
`formatAggregation `

highker commented on presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java:
```diff
@@ -341,17 +344,22 @@ public Void visitAggregation(AggregationNode node, Void context)
             return node.getSource().accept(this, context);
         }
 
+        private String printAggregation(AggregationNode.Aggregation aggregation)
```
Same

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java:
```diff
@@ -228,31 +237,99 @@ public Result apply(AggregationNode node, Captures captures, Context context)
-    private RowExpression toRowExpression(Expression expression, Rule.Context context)
+    private RowExpression toRowExpression(Expression expression, Rule.Context context, Map<NodeRef<Expression>, Type> lambdaTypes)
     {
+        if (expression instanceof LambdaExpression) {
```
separate into another helper....

wenleix commented:
"Expand FunctionCall in aggregation"

"Expand" is a bit vague, I would also explain a bit in the commit message about what is done in the commit:

```
Flatten FunctionCall in as direct fields in AggregationNode#Aggregation

Previously these information is stored in `FunctionCall`, which is an `Expression`.
This is required by following commits which will make AggregationNode to use 
`RowExpression` instead of `Expression`. 
```

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2536,10 +2537,11 @@ private AccumulatorFactory buildAccumulatorFactory(
-                    Symbol argumentSymbol = Symbol.from(argument);
-                    valueChannels.add(source.getLayout().get(argumentSymbol));
+            for (RowExpression argument : aggregation.getArguments()) {
+                checkArgument(!isExpression(argument), "cannot have untranslated expression after optimization phase");
```
nit:

wenleix commented:
Looks like Github shows the commits out of order. You might find this useful: https://github.com/isaacs/github/issues/386#issuecomment-311683282

wenleix commented:
Looks like Github shows the commits out of order. You might find this useful: https://github.com/isaacs/github/issues/386#issuecomment-311683282

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java:
```diff
@@ -146,12 +152,14 @@ public Result apply(AggregationNode parent, Captures captures, Context context)
                                 aggregation.getOrderBy(),
-                                false,
-                                Optional.of(marker)));
+                                false, Optional.of(marker)));
```
nit: new line

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java:
```diff
@@ -255,4 +264,13 @@ private PlanNode split(AggregationNode node, Context context)
                 node.getGroupIdSymbol());
     }
+
+    private boolean isLambda(RowExpression rowExpression)
```
Can this be a static method?

Another similar methods is `MultipleDistinctAggregationToMarkDistinct#toSymbol`. Should they be put in a Util class? @highker

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyCountOverConstant.java:
```diff
@@ -105,11 +114,30 @@ private boolean isCountOverConstant(AggregationNode.Aggregation aggregation, Ass
-        return argument instanceof Literal && !(argument instanceof NullLiteral);
+        return assigned instanceof Literal && !(assigned instanceof NullLiteral);
+    }
+    private static Symbol toSymbol(RowExpression rowExpression)
```
This is the same to the one in `MultipleDistinctAggregationToMarkDistinct` and `SingleDistinctAggregationToGroupBy.java` right?

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExpressionEquivalence.java:
```diff
@@ -98,6 +98,17 @@ public boolean areExpressionsEquivalent(Session session, Expression leftExpressi
         return canonicalizedLeft.equals(canonicalizedRight);
     }
 
+    public RowExpression toRowExpression(Session session, Expression expression, TypeProvider types)
```
It's not intuitive why this method is in `ExpressionEquivalence`, instead of some thing like `RowExpressionUtils`.

I understand previously  `toRowExpression(Session session, Expression expression, Map<Symbol, Integer> symbolInput, TypeProvider types)` was in this class. However, it is a private method only used by other methods in this class.  But having a public method is different.

Does it make sense to put these two `toRowExpression` into some utility class? @highker

highker replied:
I would create a new utils called `RowExpressionEquivalence`. `ExpressionEquivalence` should be marked as deprecated and `areExpressionsEquivalent` should call `RowExpressionEquivalence`

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAndExceptAsUnion.java:
```diff
@@ -248,12 +250,14 @@ private AggregationNode computeCounts(UnionNode sourceNode, List<Symbol> origina
                         Optional.empty(),
-                        false,
-                        Optional.empty()));
+                        false, Optional.empty()));
```
nit: new line

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAndExceptAsUnion.java:
```diff
@@ -292,5 +296,10 @@ private ProjectNode project(PlanNode node, List<Symbol> columns)
                     Assignments.identity(columns));
         }
+
+        private VariableReferenceExpression bigIntVariable(Symbol symbol)
```
This method is only used once. What about just inline it?

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeMixedDistinctAggregations.java:
```diff
@@ -503,7 +549,7 @@ public AggregateInfo(List<Symbol> groupBySymbols, Symbol mask, Map<Symbol, Aggre
                     .flatMap(aggregation -> aggregation.getArguments().stream())
                     .distinct()
-                    .map(Symbol::from)
+                    .map(s -> getSymbolReference(s).get())
```
nit: `symbol -> ...`, ditto for line 562.

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarAggregationToJoinRewriter.java:
```diff
@@ -177,14 +179,18 @@ private PlanNode rewriteScalarAggregation(
+                                        fromTypeSignatures(scalarAggregationSourceTypeSignatures)),
+                                BIGINT,
+                                ImmutableList.of(
+                                        new VariableReferenceExpression(nonNullableAggregationSourceSymbol.getName(),
```
I would put an line break right after `VariableReferenceExpression(`

wenleix commented on presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/AggregationFunctionMatcher.java:
```diff
@@ -60,15 +67,32 @@ public AggregationFunctionMatcher(ExpectedValueProvider<FunctionCall> callMaker)
+                Streams.zip(
+                        aggregation.getArguments().stream(),
+                        expectedCall.getArguments().stream(),
+                        (l, r) -> isEquivalent(Optional.of(r), Optional.of(l))).allMatch(Boolean::booleanValue) &&
```
instead of `l`, `r`, maybe use `actualArguments` and `expectedArguments`?

hellium01 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2536,10 +2537,11 @@ private AccumulatorFactory buildAccumulatorFactory(
+                checkArgument(!isExpression(argument), "cannot have untranslated expression after optimization phase");
+                if (!(argument instanceof LambdaDefinitionExpression)) {
+                    checkArgument(argument instanceof VariableReferenceExpression, "argument must be variable reference");
+                    valueChannels.add(source.getLayout().get(new Symbol(((VariableReferenceExpression) argument).getName())));
```
Yes, @rongrong is working on that right now.

hellium01 commented on presto-main/src/main/java/com/facebook/presto/sql/relational/SymbolToSymbolTranslator.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.util.function.Function;
+
+public class SymbolToSymbolTranslator
```
Yes, it does :) It is called `ExpressionTreeRewriter`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:
```diff
@@ -322,9 +336,21 @@ private static boolean groupsOnAllColumns(AggregationNode node, List<Symbol> col
 
+    private static OrderingScheme inlineOrderBySymbols(Map<Symbol, SymbolReference> symbolMapping, OrderingScheme orderingScheme)
+    {
+        ImmutableList.Builder orderBy = builder();
```
`ImmutableList.Builder<Symbol> orderBy = ImmutableList.builder();`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:
```diff
@@ -322,9 +336,21 @@ private static boolean groupsOnAllColumns(AggregationNode node, List<Symbol> col
         return Optional.of(new MappedAggregationInfo(aggregationOverNullRow, aggregationsSymbolMapping));
     }
 
+    private static OrderingScheme inlineOrderBySymbols(Map<Symbol, SymbolReference> symbolMapping, OrderingScheme orderingScheme)
```
Add a comment to denote this logic is from `ExpressionTreeRewriter::rewriteSortItems`?
Could you also add a test to cover this code path in `TestPushAggregationThroughOuterJoin`?

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AggregationNodeUtils.java:
```diff
@@ -0,0 +1,51 @@
+    public static Set<Symbol> extractUnique(AggregationNode.Aggregation aggregation)
+    {
+        ImmutableSet.Builder<Symbol> builder = ImmutableSet.builder();
+        aggregation.getArguments().stream().forEach(argument -> builder.addAll(SymbolsExtractor.extractAll(argument)));
```
remove `stream()`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeValidator.java:
```diff
@@ -184,10 +184,15 @@ private void checkFunctionSignature(Map<Symbol, Aggregation> aggregations)
         }
 
-        private void checkFunctionCall(Map<Symbol, Aggregation> aggregations)
+        private void checkAggregation(Map<Symbol, Aggregation> aggregations)
```
`testInvalidAggregationFunctionCall` will fail

hellium01 replied:
This test will be disabled temporarily until we converted to CallExpression.

highker commented:
@hellium01 could you extract the first 3 commits into another PR. It worth validating the correctness of those changes separately.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/DesugarAtTimeZoneRewriter.java:
```diff
@@ -57,6 +58,10 @@ public static Expression rewrite(Expression expression, Session session, Metadat
         if (expression instanceof SymbolReference) {
             return expression;
         }
+        if (expression instanceof LambdaExpression) {
```
why we need this?

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionExtractor.java:
```diff
@@ -94,15 +94,14 @@ public Void visitAggregation(AggregationNode node, ImmutableList.Builder<RowExpr
                                 .map(Symbol::toSymbolReference)
-                                .map(OriginalExpressionUtils::castToRowExpression)
-                                .forEach(context::add);
+                                .map(OriginalExpressionUtils::castToRowExpression).forEach(context::add);
```
move `forEach` to a separate line.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionSymbolInliner.java:
```diff
@@ -0,0 +1,73 @@
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.stream.Collectors.toMap;
+
+public final class RowExpressionSymbolInliner
```
Mark `ExpressionSymbolInliner` as deprecated and separate this class together with its test to a standalone commit.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionSymbolInliner.java:
```diff
@@ -0,0 +1,73 @@
+
+    public static RowExpression inlineSymbols(Map<Symbol, Symbol> mapping, RowExpression expression)
+    {
+        if (isExpression(expression)) {
```
For utils, we should always expect it to be pure `RowExpression `

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddIntermediateAggregations.java:
```diff
@@ -206,11 +216,28 @@ private PlanNode addGatheringIntermediate(AggregationNode aggregation, PlanNodeI
+                input = Symbol.from(castToExpression(argument));
+            }
+            else {
+                checkArgument(argument instanceof VariableReferenceExpression, "Non-lambda argument of final aggregation must be either SymbolReference or VariableReference");
```
rules shouldn't handle `RowExpression` for now

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneOrderByInAggregation.java:
```diff
@@ -67,12 +67,10 @@ else if (functionManager.getAggregateFunctionImplementation(aggregation.getFunct
                         Optional.empty(),
-                        aggregation.isDistinct(),
-                        aggregation.getMask()));
+                        aggregation.isDistinct(), aggregation.getMask()));
```
one param a line.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:
```diff
@@ -350,10 +357,22 @@ private static OrderingScheme inlineOrderBySymbols(Map<Symbol, SymbolReference>
+                    inputSymbols.add(new Symbol(((SymbolReference) expression).getName()));
+                }
+            }
+            else {
```
`else if`

highker commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestRowExpressionSymbolInliner.java:
```diff
@@ -0,0 +1,82 @@
+
+public class TestRowExpressionSymbolInliner
+{
+    private static final FunctionHandle TEST_FUNCTION = new FunctionHandle()
```
nit: `private static final FunctionHandle TEST_FUNCTION = () -> null;`

hellium01 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlannerUtils.java:
```diff
@@ -34,4 +45,28 @@ public static SortOrder toSortOrder(SortItem sortItem)
+        Map<Symbol, SortOrder> orderings = new LinkedHashMap<>();
+        for (SortItem item : sortItems) {
+            Symbol symbol = translator.apply(item.getSortKey());
+            // don't override existing keys, i.e. when "ORDER BY a ASC, a DESC" is specified
```
Can we leave this as a special case? We can zip it and convert but the end rows will still be similar and we end up doing unnecessary conversion:
```
      List<SortItem> translatedSortItems =
                Streams.zip(
                        orderByExpressions.stream()
                .map(subPlan::translate),
                orderBy.get().getSortItems().stream(),
                        (newSymbol, oldSortItem) -> new SortItem(newSymbol.toSymbolReference(), oldSortItem.getOrdering(), oldSortItem.getNullOrdering()))
                .collect(toImmutableList());
```

hellium01 commented:
First 3 commits were moved to #12789

hellium01 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddIntermediateAggregations.java:
```diff
@@ -206,12 +214,18 @@ private PlanNode addGatheringIntermediate(AggregationNode aggregation, PlanNodeI
-            Symbol input = getOnlyElement(extractUnique(entry.getValue()));
+            RowExpression argument = getOnlyElement(aggregation.getArguments()
+                    .stream()
+                    .filter(AddIntermediateAggregations::isNotLambda)
```
The original logic does not actually cover the lambda argument case (as it was added later).

hellium01 replied:
We don't need extractUnique here since we should not have anything like `a+1` or `a+b` in aggregation function in FINAL step (see `PushPartialAggregationThroughExchange`).

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionSymbolInliner.java:
```diff
@@ -0,0 +1,63 @@
+    @Override
+    public RowExpression rewriteLambda(LambdaDefinitionExpression node, Void context, RowExpressionTreeRewriter<Void> treeRewriter)
+    {
+        excludedNames.addAll(node.getArguments());
```
Could you add the same checkArgument logic like the one in `ExpressionSymbolInliner::rewriteLambdaExpression`?

hellium01 replied:
Will add it.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlannerUtils.java:
```diff
@@ -45,28 +44,21 @@ public static SortOrder toSortOrder(SortItem sortItem)
+    public static OrderingScheme toOrderingScheme(OrderBy orderBy)
     {
-        return toOrderingScheme(sortItems, item -> {
+        return toOrderingScheme(orderBy.getSortItems().stream().map(SortItem::getSortKey).map(item -> {
```
Let's do better code formatting
```java
return toOrderingScheme(
        orderBy.getSortItems().stream().map(SortItem::getSortKey).map(item -> { 
            checkArgument(item instanceof SymbolReference, "must be symbol reference");
            return new Symbol(((SymbolReference) item).getName()); 
        }),
        orderBy.getSortItems().stream().map(PlannerUtils::toSortOrder));

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlannerUtils.java:
```diff
@@ -45,28 +44,21 @@ public static SortOrder toSortOrder(SortItem sortItem)
         Map<Symbol, SortOrder> orderings = new LinkedHashMap<>();
-        for (SortItem item : sortItems) {
-            Symbol symbol = translator.apply(item.getSortKey());
+        Streams.forEachPair(orderingSymbols, sortOrders, (orderSymbol, sortOrder) -> {
```
I guess this is good enough
```java
forEachPair(orderingSymbols, sortOrders, orderings::putIfAbsent);

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeValidator.java:
```diff
@@ -179,11 +182,32 @@ private void checkAggregation(Map<Symbol, Aggregation> aggregations)
+                        types.get(symbol).getTypeSignature(),
+                        aggregation.getCall().getType().getTypeSignature());
+                checkArgument(functionMetadata.getArgumentTypes().size() == aggregation.getArguments().size(),
+                        "FunctionHandle input size is different from number of arguments");
```
print out the mismatched sizes

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/StatisticsAggregationPlanner.java:
```diff
@@ -80,8 +82,11 @@ public TableStatisticAggregation createStatisticsAggregation(TableStatisticsMeta
-                    ImmutableList.of(),
+                    new CallExpression(
+                            "count",
+                            functionManager.lookupFunction(count, ImmutableList.of()),
```
Use `countFunction`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/StatisticsAggregationPlanner.java:
```diff
@@ -80,8 +82,11 @@ public TableStatisticAggregation createStatisticsAggregation(TableStatisticsMeta
             }
             String count = "count";
```
remove this

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddIntermediateAggregations.java:
```diff
@@ -150,7 +153,7 @@ public Result apply(AggregationNode aggregation, Captures captures, Context cont
     }
 
-    private PlanNode addGatheringIntermediate(AggregationNode aggregation, PlanNodeIdAllocator idAllocator)
+    private PlanNode addGatheringIntermediate(AggregationNode aggregation, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
```
Previous comment is not addressed (https://github.com/prestodb/presto/pull/12710#discussion_r284573338)

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java:
```diff
@@ -155,9 +156,15 @@ public Result apply(AggregationNode aggregationNode, Captures captures, Context
+                                aggregation.getCall().getType(),
+                                aggregation.getCall().getArguments()
+                                        .stream()
+                                        .map(argument -> isExpression(argument) ? castToRowExpression(rewriter.rewrite(castToExpression(argument), context)) : argument)
```
We shouldn't have not isExpression case

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExpressionRewriteRuleSet.java:
```diff
@@ -155,9 +156,15 @@ public Result apply(AggregationNode aggregationNode, Captures captures, Context
+                                        .map(argument -> isExpression(argument) ? castToRowExpression(rewriter.rewrite(castToExpression(argument), context)) : argument)
+                                        .collect(toImmutableList())),
+                        aggregation.getFilter()
+                                .map(filter -> isExpression(filter) ? castToRowExpression(rewriter.rewrite(castToExpression(filter), context)) : filter),
```
same

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:
```diff
@@ -286,7 +288,7 @@ private static boolean groupsOnAllColumns(AggregationNode node, List<Symbol> col
             nullLiterals.add(castToRowExpression(nullLiteral));
             Symbol nullSymbol = symbolAllocator.newSymbol(nullLiteral, symbolAllocator.getTypes().get(sourceSymbol));
             nullSymbols.add(nullSymbol);
-            sourcesSymbolMappingBuilder.put(sourceSymbol, nullSymbol.toSymbolReference());
```
this change doesn't seem to be necessary

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java:
```diff
@@ -81,17 +83,25 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
     {
-        return new OrderBy(orderBy.getSortItems().stream()
-                .map(sortItem -> new SortItem(map(sortItem.getSortKey()), sortItem.getOrdering(), sortItem.getNullOrdering())).collect(Collectors.toList()));
+        // This logic is slightly different from SymbolInliner
```
different in terms of what? Complete the comment to explain

hellium01 replied:
How about?
```
// SymbolMapper inlines symbol with multiple level reference (SymbolInliner only inline single level).
```

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java:
```diff
@@ -285,4 +513,29 @@ private RowExpression toRowExpression(Expression expression, Rule.Context contex
         return SqlToRowExpressionTranslator.translate(expression, types, ImmutableMap.of(), metadata.getFunctionManager(), metadata.getTypeManager(), context.getSession(), false);
     }
+
+    private RowExpression toRowExpression(LambdaExpression expression, Rule.Context context, Map<NodeRef<Expression>, Type> lambdaTypes)
```
call it `toLambdaExpression`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java:
```diff
@@ -285,4 +513,29 @@ private RowExpression toRowExpression(Expression expression, Rule.Context contex
+        return SqlToRowExpressionTranslator.translate(expression, lambdaTypes, ImmutableMap.of(), metadata.getFunctionManager(), metadata.getTypeManager(), context.getSession(), false);
+    }
+
+    private RowExpression removeOriginalExpression(RowExpression rowExpression, Rule.Context context)
```
This function body should be delegated to the one below by having `lambdaTypes` is an empty set

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java:
```diff
@@ -285,4 +513,29 @@ private RowExpression toRowExpression(Expression expression, Rule.Context contex
+
+    private RowExpression removeOriginalExpression(RowExpression rowExpression, Rule.Context context, Map<NodeRef<Expression>, Type> lambdaTypes)
+    {
+        if (isExpression(rowExpression)) {
```
We have many places in this class doing this check, we should reuse this helper function for all use cases outside aggregation.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java:
```diff
@@ -271,6 +292,213 @@ public Result apply(ValuesNode valuesNode, Captures captures, Context context)
+        return Optional.empty();
+    }
+
+    private AggregationNode.Aggregation convertToRowExpression(AggregationNode.Aggregation aggregation, Rule.Context context)
```
Call this `translateAggregation`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java:
```diff
@@ -271,6 +292,213 @@ public Result apply(ValuesNode valuesNode, Captures captures, Context context)
+                aggregation.getMask());
+    }
+
+    private Map<NodeRef<Expression>, Type> analyzeLambdaTypes(AggregationNode.Aggregation aggregation, Rule.Context context)
```
I think the class is complicated enough to have its own test

hellium01 replied:
Will add a test on translateAggregation

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java:
```diff
@@ -271,6 +292,213 @@ public Result apply(ValuesNode valuesNode, Captures captures, Context context)
+        return ImmutableMap.of();
+    }
+
     private RowExpression toRowExpression(Expression expression, Rule.Context context)
```
This function body should be delegated to the one below by having lambdaTypes is an empty map

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java:
```diff
@@ -271,6 +292,213 @@ public Result apply(ValuesNode valuesNode, Captures captures, Context context)
+        }
+    }
+
+    private Optional<StatisticAggregations> cleanUpStatisticAggregation(StatisticAggregations statisticAggregations, Rule.Context context)
```
Call this `translateStatisticAggregation`

highker commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestTypeValidator.java:
```diff
@@ -243,32 +248,11 @@ public void testInvalidAggregationFunctionCall()
-    }
-
-    @Test(expectedExceptions = IllegalArgumentException.class, expectedExceptionsMessageRegExp = "type of symbol 'sum(_[0-9]+)?' is expected to be double, but the actual type is bigint")
-    public void testInvalidAggregationFunctionSignature()
```
Why do we remove this test?

hellium01 replied:
Rebase error, will add it back.

highker commented:
Rebase?

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2531,76 +2529,27 @@ private AccumulatorFactory buildAccumulatorFactory(
-                    Symbol argumentSymbol = Symbol.from(argument);
-                    valueChannels.add(source.getLayout().get(argumentSymbol));
+            for (RowExpression argument : aggregation.getArguments()) {
+                checkArgument(!isExpression(argument), "cannot have untranslated expression after optimization phase");
```
not necessary but fine

hellium01 replied:
Will remove it

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeMixedDistinctAggregations.java:
```diff
@@ -463,6 +479,17 @@ private AggregationNode createNonDistinctAggregation(
 
+        private static Set<Symbol> extractSymbolReferences(List<RowExpression> arguments)
+        {
+            ImmutableSet.Builder builder = builder();
```
`ImmutableSet.Builder<Symbol> builder = ImmutableSet.builder();`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeMixedDistinctAggregations.java:
```diff
@@ -463,6 +479,17 @@ private AggregationNode createNonDistinctAggregation(
+            ImmutableSet.Builder builder = builder();
+            for (RowExpression argument : arguments) {
+                if (castToExpression(argument) instanceof SymbolReference) {
+                    builder.add(new Symbol(((SymbolReference) castToExpression(argument)).getName()));
```
`Symbol.from`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeMixedDistinctAggregations.java:
```diff
@@ -473,6 +500,14 @@ private static IfExpression createIfExpression(Expression left, Expression right
+    private static Optional<Symbol> getSymbolReference(RowExpression expression)
+    {
+        if (castToExpression(expression) instanceof SymbolReference) {
+            return Optional.of(new Symbol(((SymbolReference) castToExpression(expression)).getName()));
```
`Symbol.from`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java:
```diff
@@ -81,17 +83,25 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
     {
-        return new OrderBy(orderBy.getSortItems().stream()
-                .map(sortItem -> new SortItem(map(sortItem.getSortKey()), sortItem.getOrdering(), sortItem.getNullOrdering())).collect(Collectors.toList()));
+        // This logic is slightly different from SymbolInliner
```
comment not addressed (https://github.com/prestodb/presto/pull/12710#discussion_r286711623)

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java:
```diff
@@ -81,17 +83,25 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
-        return new OrderBy(orderBy.getSortItems().stream()
-                .map(sortItem -> new SortItem(map(sortItem.getSortKey()), sortItem.getOrdering(), sortItem.getNullOrdering())).collect(Collectors.toList()));
+        // This logic is slightly different from SymbolInliner
+        ImmutableList.Builder orderBy = new ImmutableList.Builder();
```
add specialized types <...>; also use `build()` instead of `new`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java:
```diff
@@ -81,17 +83,25 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
-                .map(sortItem -> new SortItem(map(sortItem.getSortKey()), sortItem.getOrdering(), sortItem.getNullOrdering())).collect(Collectors.toList()));
+        // This logic is slightly different from SymbolInliner
+        ImmutableList.Builder orderBy = new ImmutableList.Builder();
+        ImmutableMap.Builder<Symbol, SortOrder> ordering = new ImmutableMap.Builder<>();
```
same

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java:
```diff
@@ -271,18 +294,250 @@ public Result apply(ValuesNode valuesNode, Captures captures, Context context)
+
+    private RowExpression removeOriginalExpression(RowExpression rowExpression, Session session, Map<NodeRef<Expression>, Type> types)
+    {
+        if (isExpression(rowExpression)) {
```
comment not addressed (https://github.com/prestodb/presto/pull/12710#discussion_r286713457)

hellium01 commented:
Rebased on top #12606

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionVariableInliner.java:
```diff
@@ -26,27 +26,27 @@
+    private final Map<VariableReferenceExpression, RowExpression> mapping;
 
-    private RowExpressionSymbolInliner(Map<Symbol, Symbol> mapping)
+    private RowExpressionVariableInliner(Map<VariableReferenceExpression, RowExpression> mapping)
```
nit: Why the mapping is to `RowExpression` (instead of `VariableReferenceExpression` ? :) )

hellium01 replied:
What inliner does is to replace variable reference with the value it assigned to (thus not limiting to VariableReference).

wenleix replied:
@hellium01 : I see. So it can also be an `ConstantExpression`? Basically it has more power than the original `RowExpressionSymbolInliner` now ?

hellium01 replied:
`ExpressionSymbolInliner` today also inlines Expression other than SymbolReference :)

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java:
```diff
@@ -255,4 +266,9 @@ private PlanNode split(AggregationNode node, Context context)
+
+    private static boolean isLambda(RowExpression rowExpression)
+    {
+        return castToExpression(rowExpression) instanceof LambdaExpression;
```
Curious: Why need to cast to expression? We can directly see whether it's a `LambdaDefinitionExpression` right?

hellium01 replied:
The problem is we haven't moved the translation up to this rule yet. So, it is actually a `OriginalExpression` which contains a `Expression`.

wenleix replied:
I see, thanks!

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java:
```diff
@@ -357,8 +357,11 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Set<Variab
-                    expectedInputs.addAll(extractUniqueVariables(aggregation, symbolAllocator.getTypes()));
-                    aggregation.getMask().ifPresent(expectedInputs::add);
+                    expectedInputs.addAll(
+                            extractUniqueVariables(aggregation, symbolAllocator.getTypes())
```
nit: Just do `ImmutableList.copyOf(extractUniqueVariables(aggregation, symbolAllocator.getTypes())`. ditto for line 661.

hellium01 replied:
Good catch, we don't need this conversion after the rebase. The stream/collect was needed because we need to convert to Symbol.

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java:
```diff
@@ -271,18 +271,256 @@ public Result apply(ValuesNode valuesNode, Captures captures, Context context)
+            }
+        }
+        for (RowExpression argument : aggregation.getArguments()) {
+            if (!isExpression(argument) || castToExpression(argument) instanceof LambdaExpression) {
```
Can we just do `argument instanceof LambdaDefinitionExpression` ?

hellium01 replied:
Actually, we can't here. `LambdaDefinitionExpression` is already translated `RowExpression`. Here we want to convert remaining `Expression` we hide in `OriginalExpression` into  `RowExpression`.

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeValidator.java:
```diff
@@ -170,10 +173,33 @@ private void checkFunctionSignature(Map<VariableReferenceExpression, Aggregation
+                verifyTypeSignature(
+                        variable,
+                        aggregation.getCall().getType().getTypeSignature());
+                int argumentSize = aggregation.getArguments().size();
```
Curious: Do we have corresponding check in the `Expression` world? Or they are newly enabled with this `RowExpression` migration?

hellium01 replied:
Yes, we previously cannot check argument types because we simplify don't know them.

hellium01 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlannerUtils.java:
```diff
@@ -44,18 +46,22 @@ public static SortOrder toSortOrder(SortItem sortItem)
+    {
+        return toOrderingScheme(
+                orderBy.getSortItems().stream().map(SortItem::getSortKey).map(item -> {
+                    checkArgument(item instanceof SymbolReference, "must be symbol reference");
```
This is probably the only way to use same logic to construct `OrderingScheme` :) Using for loop, we have to write the for loops multiple times.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeMixedDistinctAggregations.java:
```diff
@@ -430,12 +444,12 @@ private AggregationNode createNonDistinctAggregation(
+                    if (!duplicatedDistinctVariable.equals(distinctVariable) && extractVariables(entry.getValue().getArguments(), symbolAllocator.getTypes()).contains(distinctVariable)) {
+                        ImmutableList.Builder<RowExpression> argumentsBuilder = ImmutableList.builder();
+                        for (RowExpression argument : aggregation.getArguments()) {
+                            if (getVariable(argument, symbolAllocator.getTypes()).map(a -> a.equals(distinctVariable)).orElse(false)) {
```
I would just put the logic of `getVariable` in this loop. `getVariable` is too generic name to make ambiguity.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeMixedDistinctAggregations.java:
```diff
@@ -513,8 +553,7 @@ public AggregateInfo(List<VariableReferenceExpression> groupByVariables, Variabl
                     .distinct()
-                    .map(Symbol::from)
-                    .map(symbol -> new VariableReferenceExpression(symbol.getName(), types.get(symbol)))
+                    .map(symbol -> getVariable(symbol, types).get())
```
inline `getVariable`

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java:
```diff
@@ -358,7 +358,7 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Set<Variab
                     Aggregation aggregation = entry.getValue();
                     expectedInputs.addAll(extractUniqueVariables(aggregation, symbolAllocator.getTypes()));
-                    aggregation.getMask().ifPresent(expectedInputs::add);
+                    aggregation.getMask().ifPresent(mask -> expectedInputs.add(mask));
```
no need to change

hellium01 replied:
Could be a result of rebase (Symbol/Variable conversion is now removed).

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java:
```diff
@@ -111,17 +113,25 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
     {
-        return new OrderingScheme(orderingScheme.getOrderBy().stream().map(this::map).collect(toImmutableList()),
-                orderingScheme.getOrderings().entrySet().stream().collect(toMap(entry -> map(entry.getKey()), entry -> entry.getValue())));
+        if (isExpression(value)) {
```
let's move this part of logic to the caller. The caller should decide which method to use.

hellium01 replied:
SymbolMapper will be used to in multiple optimization rules. As we are pushing the clean up up, we will need this function to handle both Expression/RowExpression anyway. I would prefer we just remove isExpression later once everything is cleaned up.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java:
```diff
@@ -79,16 +81,21 @@ public Parts createPartialAggregations(SymbolAllocator symbolAllocator, Function
                     originalAggregation.getOrderBy(),
-                    originalAggregation.isDistinct(),
-                    originalAggregation.getMask()));
+                    originalAggregation.isDistinct(), originalAggregation.getMask()));
```
new line "originalAggregation.getMask()"

highker commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTranslateExpression.java:
```diff
@@ -0,0 +1,160 @@
+import static com.facebook.presto.testing.assertions.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+
+public class TestTranslateExpression
```
add "s" at the end.

highker commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTranslateExpression.java:
```diff
@@ -0,0 +1,160 @@
+        return call("MULTIPLY", FUNCTION_RESOLUTION.arithmeticFunction(OperatorType.MULTIPLY, left.getType(), right.getType()), left.getType(), ImmutableList.of(left, right));
+    }
+
+    private static boolean hasUntranslated(AggregationNode.Aggregation aggregation)
```
`isUntranslated`

highker commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTranslateExpression.java:
```diff
@@ -0,0 +1,160 @@
+        assertFalse(hasUntranslated(translated));
+    }
+
+    private CallExpression gt(RowExpression left, RowExpression right)
```
`greaterThan`

hellium01 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AggregationNodeUtils.java:
```diff
@@ -35,32 +40,33 @@ private AggregationNodeUtils() {}
-        return builder.build();
-    }
-
     public static Set<VariableReferenceExpression> extractUniqueVariables(AggregationNode.Aggregation aggregation, TypeProvider types)
```
You mean rename Aggregation to AggregationFunction?

highker replied:
I mean rename the function; just check the one in `WindowNodeUtil `

hellium01 replied:
Will change to `extractAggregationUniqueVariables`

mbasmanova commented:
@hellium01 Yi, I'm seeing `Streaming aggregation with input not grouped on the grouping keys` planning error for the following query. Could it be related to this change?

```
SELECT
      linenumber,
      'xxx'
  FROM
  (
      (SELECT orderkey, linenumber FROM lineitem)
      UNION
      (SELECT orderkey, linenumber FROM lineitem)
  ) WHERE orderkey = 1 
  GROUP BY
      1;



java.lang.IllegalArgumentException: Streaming aggregation with input not grouped on the grouping keys

    at com.google.common.base.Preconditions.checkArgument(Preconditions.java:135)
    at com.facebook.presto.sql.planner.sanity.ValidateStreamingAggregations$Visitor.visitAggregation(ValidateStreamingAggregations.java:89)
    at com.facebook.presto.sql.planner.sanity.ValidateStreamingAggregations$Visitor.visitAggregation(ValidateStreamingAggregations.java:52)
    at com.facebook.presto.sql.planner.plan.AggregationNode.accept(AggregationNode.java:200)
    at com.facebook.presto.sql.planner.plan.InternalPlanNode.accept(InternalPlanNode.java:31)
    at com.facebook.presto.sql.planner.sanity.ValidateStreamingAggregations$Visitor.lambda$visitPlan$0(ValidateStreamingAggregations.java:73)
```
